base/loading.jl

```diff
@@ -181,3 +216,48 @@ function evalfile(path::AbstractString, args::Vector{UTF8String}=UTF8String[])
                      :(include($path))))
 end
 evalfile(path::AbstractString, args::Vector) = evalfile(path, UTF8String[args...])
+
+create_expr_cache(m::Box) = create_expr_cache(m.contents::Expr)
+function create_expr_cache(m::Expr)
+    assert(m.head === :module)
+    typeassert(m.args[2], Symbol)
+    cachepath = LOAD_CACHE_PATH[1]
+    if !isdir(cachepath)
+        mkdir(cachepath)
+    end
+    code_object = """
+        while !eof(STDIN)
+            eval(Main, deserialize(STDIN))
+        end
+        """
+    io, pobj = open(detach(setenv(`$(julia_cmd()) --build $cachepath --startup-file=no --history-file=no -E $code_object`,["JULIA_HOME=$JULIA_HOME","HOME=$(homedir())"])), "w")
+    serialize(io, quote
+        empty!(Base.LOAD_PATH)
+        append!(Base.LOAD_PATH, $LOAD_PATH)
+        empty!(Base.LOAD_CACHE_PATH)
+        append!(Base.LOAD_CACHE_PATH, $LOAD_CACHE_PATH)
+        empty!(Base.DL_LOAD_PATH)
+        append!(Base.DL_LOAD_PATH, $DL_LOAD_PATH)
+    end)
+    source = source_path(nothing)
+    if source !== nothing
+        serialize(io, quote
+            task_local_storage()[:SOURCE_PATH] = $(source)
+        end)
+    end
+    serialize(io, m)
+    if source !== nothing
+        serialize(io, quote
+            delete!(task_local_storage(), :SOURCE_PATH)
+        end)
+    end
+    close(io)
+    wait(pobj)
+    if !_require_from_cache(string(m.args[2]))
+        error("Warning: @cacheable module failed to define a module")
+    end
+end
+
+macro cacheable(m)
+    :(create_expr_cache($(Box(m))))
```

Comment:
Well, we should do something else, or maybe change the lowering code. It can't be that this is the only use case for such a thing, or that `Box` is uniquely suited to solving this problem.